File: Chapter- 15/union_find.cpp
#include <bits/stdc++.h>
using namespace std;

class Edge{
public:
	int source, dest;
};

class Graph{
public:
	int V, E;
	Edge* edge;
};

Graph* create(int V, int E){
	Graph* graph = new Graph();
	graph->V = V;
	graph->E = E;

	graph->edge = new Edge[graph->E * sizeof(Edge)];

	return graph;
}

int find_parent(int parent[], int i){
	if (parent[i] == -1)
		return i;
	return find_parent(parent, parent[i]);
}

void find_union(int parent[], int x, int y){
	parent[x] = y;
}

int detect_Cycle(Graph* graph){
	int* parent = new int[graph->V * sizeof(int)];

	memset(parent, -1, sizeof(int) * graph->V);

	for (int i = 0; i < graph->E; ++i) {
		int x = find_parent(parent, graph->edge[i].source);
		int y = find_parent(parent, graph->edge[i].dest);

		if (x == y)
			return 1;

		find_union(parent, x, y);
	}
	return 0;
}

int main(){
	int V = 5, E = 6;
	Graph* graph = create(V, E);

	graph->edge[0].source = 0;
	graph->edge[0].dest = 1;

	graph->edge[1].source = 1;
	graph->edge[1].dest = 2;

	graph->edge[2].source = 2;
	graph->edge[2].dest = 3;

    graph->edge[3].source = 3;
	graph->edge[3].dest = 4;

	graph->edge[4].source = 4;
	graph->edge[4].dest = 5;

	graph->edge[5].source = 5;
	graph->edge[5].dest = 1;

	if (detect_Cycle(graph))
		cout << " graph contains cycle ";
	else
		cout << " graph doesn't contain cycle";

	return 0;
}

File: Chapter- 15/union_find.py
from collections import defaultdict

class Graph:

    def __init__(self, vertex):
        self.V = vertex
        self.graph = defaultdict(list)

    def Add_edge(self, u, v):
        self.graph[u].append(v)

    def parent(self, parent, i):
        if parent[i] == -1:
            return i
        if parent[i] != -1:

            return self.parent(parent, parent[i])

    def union(self, parent, x, y):
        parent[x] = y

    def Check_cyclic(self):

        parent = [-1] * self.V

        for i in self.graph:
            for j in self.graph[i]:
                x = self.parent(parent, i)
                y = self.parent(parent, j)
                if x == y:
                    return True
                self.union(parent, x, y)


G = Graph(5)
G.Add_edge(0, 1)
G.Add_edge(0, 2)
G.Add_edge(1, 2)
G.Add_edge(2, 3)
G.Add_edge(3, 4)
G.Add_edge(4, 5)
G.Add_edge(5, 1)

if G.Check_cyclic():
    print(" graph contains cycle")
else:
    print(" graph doesn't contain cycle ")